tsdb/head_test.go

```diff
@@ -3245,3 +3245,102 @@ func TestReplayAfterMmapReplayError(t *testing.T) {
 
 	require.NoError(t, h.Close())
 }
+
+// https://github.com/prometheus/prometheus/issues/10043
+func TestOutOfSequenceMMappeDChunks(t *testing.T) {
+	dir := t.TempDir()
+	var h *Head
+	var err error
+
+	openHead := func() {
+		wlog, err := wal.NewSize(nil, nil, filepath.Join(dir, "wal"), 32768, false)
+		require.NoError(t, err)
+
+		opts := DefaultHeadOptions()
+		opts.ChunkRange = DefaultBlockDuration
+		opts.ChunkDirRoot = dir
+		opts.EnableMemorySnapshotOnShutdown = true
+		opts.MaxExemplars.Store(config.DefaultExemplarsConfig.MaxExemplars)
+
+		h, err = NewHead(nil, nil, wlog, opts, nil)
+		require.NoError(t, err)
+		require.NoError(t, h.Init(0))
+	}
+
+	openHead()
+
+	itvl := int64(15 * time.Second / time.Millisecond)
+	lastTs := int64(0)
+	lbls := labels.FromStrings("__name__", "testing", "foo", "bar")
+	var expSamples []tsdbutil.Sample
+	addSamples := func(numSamples int) {
+		app := h.Appender(context.Background())
+		var ref storage.SeriesRef
+		for i := 0; i < numSamples; i++ {
+			ref, err = app.Append(ref, lbls, lastTs, float64(lastTs))
+			expSamples = append(expSamples, sample{t: lastTs, v: float64(lastTs)})
+			require.NoError(t, err)
+			lastTs += itvl
+			if i%10 == 0 {
+				require.NoError(t, app.Commit())
+				app = h.Appender(context.Background())
+			}
+		}
+		require.NoError(t, app.Commit())
+	}
+
+	// Creating multiple m-map files.
+	for i := 0; i < 5; i++ {
+		addSamples(250)
+		require.NoError(t, h.Close())
+		if i != 4 {
+			// Don't open head for the last iteration.
+			openHead()
+		}
+	}
+
+	chunksDir := filepath.Join(dir, "chunks_head")
+	files, err := ioutil.ReadDir(chunksDir)
+	require.Equal(t, 5, len(files))
+
+	// Make chunks out of sequence by moving 000003 samples after 000004.
+	err = os.Rename(filepath.Join(chunksDir, "000003"), filepath.Join(chunksDir, "000003.old"))
+	require.NoError(t, err)
+	err = os.Rename(filepath.Join(chunksDir, "000004"), filepath.Join(chunksDir, "000003"))
+	require.NoError(t, err)
+	err = os.Rename(filepath.Join(chunksDir, "000003.old"), filepath.Join(chunksDir, "000004"))
+	require.NoError(t, err)
+
+	openHead()
+	require.Equal(t, 1.0, prom_testutil.ToFloat64(h.metrics.mmapChunkCorruptionTotal))
+
+	// There should be less m-map files due to corruption.
+	files, err = ioutil.ReadDir(chunksDir)
+	require.Equal(t, 4, len(files))
+
+	// Make sure we can add new chunks after recovery.
+	addSamples(250)
+	require.NoError(t, h.Close())
+	openHead()
+
+	// Update the expected Sample to remove the deleted samples.
+	start, finish := 0, 0
+	for idx, v := range expSamples {
+		if v.T() == int64(7200000) {
+			start = idx
+		}
+
+		if v.T() == int64(9795000) {
+			finish = idx
+		}
```

Comment:
**note:** I found out these timestamps by manually checking the extra `(t,v)` we were getting in the [assertion](https://github.com/prometheus/prometheus/pull/10406/files#diff-7c1dfea5c568897587621510c04d322f908a8ed40751e592d63d2f8422b6a78dR3343). 

I'm not sure if there is a better way to find this out and make it a bit more "dynamic" where we don't have to know "magic" timestamps.